### PR TITLE
[spring] Resolve regression on RequestParam for non-objects

### PR DESCRIPTION
**Pull request by jimschubert**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Proposed fix for #3248, which sets model state if a property is object && model. Moved isModel assignment after the protected helper method, which already made an assignment to isModel it a not-very-robust way.

### CODE REVIEW

**Comment by peyerroger** [comment]
@jimschubert 
Hi Jim,
I don't think the problem is the  following line.

property.isModel = ModelUtils.isObjectSchema(p) && ModelUtils.isModel(p);

The fix introduced in the commit that broke the spring generator seems very invasive. I would rather just remove it and use isPrimitiveType instead. This would be a much simpler fix without the risk of other regressions. Let me prepare a proposal. I already have quite some extensive tests that covers most of the use cases. 

Cheers,
Roger

**Comment by jimschubert** [comment]
@peyerroger I think `isModel` in the template is "correct".

For example, in your PR #3857, handling of collections has changed:

```
-    default ResponseEntity<List<Pet>> findPetsByTags(@NotNull @ApiParam(value = "Tags to filter by", required = true) @Valid @RequestParam(value = "tags", required = true) List<String> tags) {
+    default ResponseEntity<List<Pet>> findPetsByTags(@NotNull @ApiParam(value = "Tags to filter by", required = true) @Valid List<String> tags) {
```

Also, your fix results in `isModel` being incorrectly set (`setNonArrayMapProperty` considers anything not a primitive to be a model, which isn't the case). While it might work for built-in templates, it would be introducing a bug for anyone who creates a custom template and expects `isModel` to work as intended.

**Comment by jimschubert** [comment]
Go technical committee (@antihax (2017/11) @bvwells (2017/12) @grokify (2018/07) @kemokemo (2018/09) @bkabrda (2019/07)):

The changes I've made in this PR have resulted in the following Go client changes and seems to unexpected fix a regression introduced after 4.0.0 in the go client ([see 4.0.0](https://github.com/OpenAPITools/openapi-generator/blob/37442733120ca60a5a28e765e377871da20d37d0/samples/client/petstore/go/go-petstore/api_fake.go#L835))…

```diff
diff --git a/samples/client/petstore/go/go-petstore-withXml/api_fake.go b/samples/client/petstore/go/go-petstore-withXml/api_fake.go
index 08859ad9f7..c08e9e995e 100644
--- a/samples/client/petstore/go/go-petstore-withXml/api_fake.go
+++ b/samples/client/petstore/go/go-petstore-withXml/api_fake.go
@@ -835,11 +835,7 @@ func (a *FakeApiService) TestEndpointParameters(ctx _context.Context, number flo
 		localVarFormParams.Add("date", parameterToString(localVarOptionals.Date.Value(), ""))
 	}
 	if localVarOptionals != nil && localVarOptionals.DateTime.IsSet() {
-		paramJson, err := parameterToJson(localVarOptionals.DateTime.Value())
-		if err != nil {
-			return nil, err
-		}
-		localVarFormParams.Add("dateTime", paramJson)
+		localVarFormParams.Add("dateTime", parameterToString(localVarOptionals.DateTime.Value(), ""))
 	}
 	if localVarOptionals != nil && localVarOptionals.Password.IsSet() {
 		localVarFormParams.Add("password", parameterToString(localVarOptionals.Password.Value(), ""))
diff --git a/samples/client/petstore/go/go-petstore/api_fake.go b/samples/client/petstore/go/go-petstore/api_fake.go
index 392ab5c46e..731d5c00f4 100644
--- a/samples/client/petstore/go/go-petstore/api_fake.go
+++ b/samples/client/petstore/go/go-petstore/api_fake.go
@@ -834,11 +834,7 @@ func (a *FakeApiService) TestEndpointParameters(ctx _context.Context, number flo
 		localVarFormParams.Add("date", parameterToString(localVarOptionals.Date.Value(), ""))
 	}
 	if localVarOptionals != nil && localVarOptionals.DateTime.IsSet() {
-		paramJson, err := parameterToJson(localVarOptionals.DateTime.Value())
-		if err != nil {
-			return nil, err
-		}
-		localVarFormParams.Add("dateTime", paramJson)
+		localVarFormParams.Add("dateTime", parameterToString(localVarOptionals.DateTime.Value(), ""))
 	}
 	if localVarOptionals != nil && localVarOptionals.Password.IsSet() {
 		localVarFormParams.Add("password", parameterToString(localVarOptionals.Password.Value(), ""))
```

I just wanted to check that the resulting change is expected. That is, am I correct that the go client hasn't begun expecting dates to be converted via `parameterToJson`?

**Comment by peyerroger** [comment]
@jimschubert 

Hi Jim,

I of course trust you if you say isModel is correct. I will close my PR as soon as yours is integrated. Could you do me a favour and add the test introduced in my PR. This will ensure that my use cases are working properly and prevents potential regressions in the future.

Test:

```
    @Test
    public void doGenerateRequestParamForSimpleParam() throws IOException {
        File output = Files.createTempDirectory("test").toFile().getCanonicalFile();
        output.deleteOnExit();
        String outputPath = output.getAbsolutePath().replace('\\', '/');

        OpenAPI openAPI = new OpenAPIParser()
            .readLocation("src/test/resources/3_0/issue_3248.yaml", null, new ParseOptions()).getOpenAPI();

        SpringCodegen codegen = new SpringCodegen();
        codegen.setOutputDir(output.getAbsolutePath());
        codegen.additionalProperties().put(CXFServerFeatures.LOAD_TEST_DATA_FROM_FILE, "true");

        ClientOptInput input = new ClientOptInput();
        input.openAPI(openAPI);
        input.config(codegen);

        MockDefaultGenerator generator = new MockDefaultGenerator();
        generator.opts(input).generate();

        checkFileContains(generator, outputPath + "/src/main/java/org/openapitools/api/MonkeysApi.java",  "@RequestParam");
        checkFileContains(generator, outputPath + "/src/main/java/org/openapitools/api/ElephantsApi.java",  "@RequestParam");
        checkFileContains(generator, outputPath + "/src/main/java/org/openapitools/api/ZebrasApi.java",  "@RequestParam");
        checkFileContains(generator, outputPath + "/src/main/java/org/openapitools/api/BearsApi.java",  "@RequestParam");
        checkFileContains(generator, outputPath + "/src/main/java/org/openapitools/api/CamelsApi.java",  "@RequestParam");
        checkFileContains(generator, outputPath + "/src/main/java/org/openapitools/api/PandasApi.java",  "@RequestParam");
        checkFileContains(generator, outputPath + "/src/main/java/org/openapitools/api/CrocodilesApi.java",  "@RequestParam");
        checkFileContains(generator, outputPath + "/src/main/java/org/openapitools/api/PolarBearsApi.java",  "@RequestParam");

    }

    private void checkFileContains(MockDefaultGenerator generator, String path, String... lines) {
        String file = generator.getFiles().get(path);
        assertNotNull(file);
        for (String line : lines)
            assertTrue(file.contains(line));
    }
```

Example API:
```
openapi: 3.0.0
servers:
  - url: 'localhost:8080'
info:
  version: 1.0.0
  title: OpenAPI Zoo
  license:
    name: Apache-2.0
    url: 'http://www.apache.org/licenses/LICENSE-2.0.html'
paths:
  /monkeys:
    get:
      operationId: getMonkeys
      parameters:
        - $ref: '#/components/parameters/refDate'
  /elephants:
    get:
      operationId: getElephants
      parameters:
        - in: query
          name: someDate
          required: true
          schema:
            type: string
            format: date
  /girafes:
    get:
      operationId: getGirafes
      parameters:
        - $ref: '#/components/parameters/refStatus'
  /zebras:
    get:
      operationId: getZebras
      parameters:
        - in: query
          name: status
          required: true
          schema:
            type: integer
            enum: [0,1]
            default: 0
  /bears:
    get:
      operationId: getBears
      parameters:
        - $ref: '#/components/parameters/refCondition'
  /camels:
    get:
      operationId: getCamels
      parameters:
        - in: query
          name: condition
          required: true
          schema:
            type: string
            enum:
              - sleeping
              - awake
  /pandas:
    get:
      operationId: getPandas
      parameters:
        - $ref: '#/components/parameters/refName'
  /crocodiles:
    get:
      operationId: getCrocodiles
      parameters:
        - in: query
          name: name
          required: true
          schema:
            type: string
  /polarBears:
    get:
      operationId: getPolarBears
      parameters:
        - $ref: '#/components/parameters/refAge'
  /birds:
    get:
      operationId: getBirds
      parameters:
        - in: query
          name: age
          required: true
          schema:
            type: integer
components:
  parameters:
    refDate:
      in: query
      name: refDate
      required: true
      schema:
        type: string
        format: date
    refStatus:
      in: query
      name: refStatus
      required: true
      schema:
        type: integer
        enum: [0,1]
        default: 0
    refCondition:
      in: query
      name: refCondition
      schema:
        type: string
        enum:
          - sleeping
          - awake
    refName:
      in: query
      name: refName
      schema:
        type: string
    refAge:
      in: query
      name: refAge
      schema:
        type: integer
```

**Comment by jimschubert** [comment]
@peyerroger  I have one remaining issue, in that ocaml regression tests are failing. I'll need to resolve this before the PR can be merged.

**Comment by jimschubert** [comment]
@OpenAPITools/generator-core-team 

This fixes #3248 which has been open since June. I've had to make changes to DefaultCodegen which sets the `isModel` property on parameters in a more meaningful way. The changes in DefaultCodegen may also resolve a handful of currently open bugs for the Go client (such as #2980). See go client output changes in this PR for a simple example of the regression.  Also labeling as Go Client due to that regression fix, which I compared against 4.0.0 release.

**Comment by jimschubert** [comment]
@t2y can you check that this resolves all of your issue in #3248?

**Comment by t2y** [comment]
@jimschubert OK. I will!

**Comment by t2y** [comment]
@jimschubert I confirmed that this change resolves my issue using my production spec. The generated source by v4.0.1 and v4.1.2-snapshot (this pr) are almost same (handling defaultValue has a diff, but it's not an issue. I think it's improved). Thank you!

**Comment by jimschubert** [comment]
Cool. Thanks for verifying! I'll wait for another core team member to review before merge, since there's a change in DefaultCodegen that affects all generators (even if it's in a --hopefully-- positive way).

**Comment by jimschubert** [comment]
CC Java technical committee:
@bbdouglas @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog

**Comment by t2y** [comment]
I hope this fix will be merged into 4.1.2. 🙏

**Comment by jimschubert** [comment]
@wing328 

> I saw some changes to the Go client and will need to review on that to understand more about the impact.

Please compare changes in this PR to version 4.0.0 for the go client output: https://github.com/OpenAPITools/openapi-generator/blob/37442733120ca60a5a28e765e377871da20d37d0/samples/client/petstore/go/go-petstore/api_fake.go#L836

File: samples/client/petstore/go/go-petstore/api_fake.go [hunk 1]
```diff
@@ -834,11 +834,7 @@ func (a *FakeApiService) TestEndpointParameters(ctx _context.Context, number flo
 		localVarFormParams.Add("date", parameterToString(localVarOptionals.Date.Value(), ""))
 	}
 	if localVarOptionals != nil && localVarOptionals.DateTime.IsSet() {
-		paramJson, err := parameterToJson(localVarOptionals.DateTime.Value())
-		if err != nil {
-			return nil, err
-		}
-		localVarFormParams.Add("dateTime", paramJson)
+		localVarFormParams.Add("dateTime", parameterToString(localVarOptionals.DateTime.Value(), ""))
```
Comment (wing328) [inline samples/client/petstore/go/go-petstore/api_fake.go]: The fix is good:
```
        - name: date
          type: string
          format: date
          in: formData
          description: None
        - name: dateTime
          type: string
          format: date-time
          in: formData
          description: None
```
so there shouldn't be any nil check as the parameter is not required.

**Comment by t2y** [comment]
@jimschubert @wing328 Thank you for detail release cycle. I don't mind this fix won't include in 4.1.2. I will verify with 4.1.3-SNAPSHOT until 4.1.3 release. I appreciate you helping this issue.

**Pull request merged**